Added retrofit2 module.

bfg opened the pull request:
This PR adds [Retrofit2](http://square.github.io/retrofit/) async-http-client extras module.

[Project Lombok](https://projectlombok.org/features/all) is used to evade boilerplate.

bfg commented:
Seems like unrelated tests are failing.

```
  NettyReactiveStreamsTest>ReactiveStreamsTest.testConnectionDoesNotGetClosed:95 Invalid response byte at position 12000 expected [111] but found [45]

Tests run: 503, Failures: 2, Errors: 0, Skipped: 0

[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] Asynchronous Http Client Project ................... SUCCESS [  0.864 s]
[INFO] Asynchronous Http Client Netty Utils ............... SUCCESS [  2.335 s]
[INFO] Asynchronous Http Client ........................... FAILURE [03:31 min]
```

slandelle commented:
> This PR adds Retrofit2 async-http-client extras module.

Thanks! Being curious, why do you use AHC with Retrofit? I mean I would expect anyone going with Retrofit to go with the full Square stack and use OkHttp instead.

> Project Lombok is used to evade boilerplate.

Correct me if I'm wrong, but such code generation causes stacktraces to not be aligned with source code lines, right? If so, I'm fine with using Lombock for this module, but it won't propagate to main one.

> Seems like unrelated tests are failing.

Yeah, that's #1380.

slandelle commented on extras/retrofit2/src/main/java/org/asynchttpclient/extras/retrofit/AsyncHttpClientCall.java:
```diff
@@ -0,0 +1,305 @@
+    private void throwAlreadyExecuted() {
+        throw new IllegalStateException("This call has already been executed.");
+    }
+}
```
missing line feed

slandelle commented on extras/retrofit2/src/main/java/org/asynchttpclient/extras/retrofit/AsyncHttpClientCallFactory.java:
```diff
@@ -0,0 +1,70 @@
+                .request(request)
+                .build();
+    }
+}
```
missing line feed

slandelle commented on extras/retrofit2/src/test/java/org/asynchttpclient/extras/retrofit/AsyncHttpClientCallFactoryTest.java:
```diff
@@ -0,0 +1,73 @@
+        };
+    }
+
+}
```
missing line feed

slandelle commented on extras/retrofit2/src/test/java/org/asynchttpclient/extras/retrofit/AsyncHttpClientCallFactoryTest.java:
```diff
@@ -0,0 +1,73 @@
+        return e -> {
+        };
+    }
+
```
kill empty line

slandelle commented on extras/retrofit2/src/test/java/org/asynchttpclient/extras/retrofit/AsyncHttpClientCallTest.java:
```diff
@@ -0,0 +1,216 @@
+    private void doThrow(String message) {
+        throw new RuntimeException(message);
+    }
+}
```
missing line feed

slandelle commented on extras/retrofit2/src/test/java/org/asynchttpclient/extras/retrofit/AsyncHttpClientCallFactoryTest.java:
```diff
@@ -0,0 +1,73 @@
+                .build();
+
+        // then
+        Assert.assertTrue(factory.getHttpClient() == httpClient);
```
Use Assert static imports

slandelle replied:
Use assertEquals (everywhere it makes sense)

slandelle commented on extras/retrofit2/src/test/java/org/asynchttpclient/extras/retrofit/AsyncHttpRetrofitIntegrationTest.java:
```diff
@@ -0,0 +1,175 @@
+        httpClient.close();
+    }
+
+    @Test(enabled = false)
```
Why are all the tests in this class disabled?

bfg replied:
They're using Github API, which is rate-limited and github may not be up all the time. Do you want me to convert this test so that it would start `wiremock` or  similar  webserver on local host and perform http calls against it?

slandelle replied:
Got it. Please have a look at the `org.asynchttpclient.testserver` package. Then, we might need some maven build refactoring to the test jar is visible from extra modules.

slandelle commented on extras/retrofit2/pom.xml:
```diff
@@ -0,0 +1,63 @@
+
+    <properties>
+      <retrofit2.version>2.3.0</retrofit2.version>
+      <lombok.version>1.16.16</lombok.version>
```
Actually, I'd rather not use lombok in there so coding style is consistent.

bfg replied:
I can convert all the code to normal java code, but frankly i don't think it's worth it - some other open source projects, namely spring-cloud started using lombok 2-3 years ago and it seem to work for them very well.

So, do you want me to de-lombok the code?

slandelle replied:
You can leave Lombok there. But I don't think it will ever make it into main module.

bfg commented:
I can convert all the code to normal java code, but frankly i don't think it's worth it - some other open source projects, namely `spring-cloud` started using lombok 2-3 years ago and it seem to work for them very well.

So, do you want me to de-lombok the code?

bfg commented:
Ok, @slandelle i've addressed all your code review comments and added the following functionality:

* integration test suite now uses `org.asynchttpclient.testserver` test http server
* `Call` and `Call.Factory` now allow multiple consumers to be attached for specific event types, allowing easy customization

Please review again. If you're okay with it, i will squash everything into one commit and resubmit.

slandelle commented:
Thanks! Note that I fixed a NPE issue (see next commit).

Cheers!

bfg commented:
Whoa! Thanks for cooperation! :-)
